File: ruby/p3.rb
def gen(n)
    yield 2
    yield 3
    max = Math.sqrt(n).to_int
    for i in 1..((n+1)/6).to_int
        [6*i-1, 6*i+1].each do |rslt|
            if rslt > max
                return
            end
            yield rslt
        end
    end
end

def find_largest_prime_factor(num)
    rslt = n = num
    gen(num) do |i|
        if n <= 1
            return rslt
        end
        while n % i == 0 do
            n /= i
            rslt = i
        end
    end

    return rslt
end

puts find_largest_prime_factor(600851475143)

File: js/p3.js
function check(n, cb) {
    if (cb(2)) return;
    if (cb(3)) return;
    var max = Math.ceil(Math.sqrt(n));
    for (var i = 1; i < Math.ceil((n + 1) / 6); i++) {
        var li = [6*i - 1, 6*i + 1];
        for (var k in li) {
            if (li[k] > max) {
                return;
            }
            if(cb(li[k])) return;
        }
    }
}

function find_largest_prime_factor(num) {
    var rslt = num, n = num;

    while (n > 1) {
        check(num, function(i) {
            while (n % i == 0) {
                n /= i;
                rslt = i;
            }
        });
    }

    return rslt;
}

console.log(find_largest_prime_factor(600851475143));
